### ISSUE TITLE
zone name 'Etc/GMT+7' leading to invalid date times

### ISSUE THREAD

**Issue by SamMatthewsIsACommonName** [moment/luxon#150]
Hi,
We've had a situation where a user's timezone information (from DateTime.local().zoneName) has been 'Etc/GMT+7'.
This has lead to an invalid DateTime being thrown when setting the zone to this using the DateTime.fromObject({ .... zone: thisZone  }) method.
I'm a bit confused because presumably the DateTime was initially valid to get the zoneName? Is that right?
Thanks
  
Edit: this invalid datetime happens on our server, but not on the client side using the same logic to format the date.
I imagine this has something to do with our node server not having access to Intl library? Is there a way to manually pull this in and configure luxon to look for it essentially?

**Comment by icambron** [comment]
Luxon uses Intl.DateTimeFormat to handle zones and unfortunately "Etc/GMT+7" is a zone with inconsistent platform support. In Chrome and Node:

```js
new Intl.DateTimeFormat('en', {timeZone: 'Etc/GMT+7'}).format()
VM250:1 Uncaught RangeError: Expected Area/Location(/Location)* for time zone, got Etc/GMT+7
    at new DateTimeFormat (native)
    at <anonymous>:1:1
```

I think the parser is choking on the "+". In FF it works just fine. Additionally, I bet even Chrome is able to output that zone when the OS is configured to it (though I haven't tested that), just not take it as input. So either that or a different browser explain the asymmetry you're seeing.

The upshot of all this is that V8 needs a fix. I *think* this is probably the best Chromium issue for tracking this and related zone parsing issues: https://bugs.chromium.org/p/chromium/issues/detail?id=364374. In the meantime there isn't much we can do from the Luxon side.

**Comment by SamMatthewsIsACommonName** [comment]
Aah I see, thanks for pointing to the issue. I guess all that can be done is to implement some sort of fallback mechanism and cross fingers that they fix it soon!
Thanks

**Comment by icambron** [comment]
Yeah. One trick would be to check it explicitly and then use Luxon's direct support for "UTC+7" as a zone identifier. I guess I could even build that in.

**Comment by SamMatthewsIsACommonName** [comment]
aah I see, so if it's invalid then query the string for the Etc/, then find the '+' amount and construct a 'UTC+X' value from that?

**Comment by icambron** [comment]
Right. Fix goes somewhere in here: https://github.com/moment/luxon/blob/master/src/impl/util.js#L182. I'd take a PR.

**Comment by SamMatthewsIsACommonName** [comment]
Ok I'll play around with it now. Thanks

**Comment by icambron** [comment]
The only downside of this approach is that for browsers that *can* handle `Etc/GMT[+-]\d+`, they won't get internationalized formatting on the name of the zone. I think making the zone working consistently is worth that price though.

**Comment by SamMatthewsIsACommonName** [comment]
In that case would it be possible to put this fix AFTER the datetime is constructed and before it returns 'invalid datetime'? So a fallback rather than a pre-emptive check?

**Comment by icambron** [comment]
Meaning like in a try/catch? It would need to go in a few different places. I think it's simpler to do it preemptively.

**Comment by SamMatthewsIsACommonName** [comment]
Yeah. OK yeah I can imagine that getting quite messy. Will get onto it now thanks

**Comment by SamMatthewsIsACommonName** [comment]
Ok I'm an idiot and can't seem to push to the branch I made, so here's the changes I made:
https://gist.github.com/SamMatthewsIsACommonName/93b8724422c6ed42cdfc1eb2f65e7e68
I added the function compensateForEtcZone and replaced input with a compensated value if it's a string. It's probably pretty horrible from every aspect so feel free to make it however you want (I didn't want to use regexp because I rightly or wrongly believe it can introduce security vulnerabilities)
But it worked:
<img width="908" alt="screen shot 2018-01-10 at 12 45 53 pm" src="https://user-images.githubusercontent.com/21329063/34753792-b2431678-f605-11e7-932e-96a20c342584.png">

**Comment by icambron** [comment]
Thanks for taking a shot at it. There are a couple problems there:

 * There are - offsets as well as + ones
 * There are Etc/* zones that are the GMT-with-offset ones

Regexes only cause vulnerabilities if you're careless.

I went ahead and fixed it here: 3b95def. I'll release it soon.

**Comment by SamMatthewsIsACommonName** [comment]
Gotcha thanks!
